### PR TITLE
Fixes #1749 Add GitHub Actions example: Resolves dstackai issue #1749 

### LINKED ISSUE #1749: [Examples] Add GitHub Actions example

### PR DESCRIPTION
**Pull request by Tob-iee**
Hi guys, I created this example for the implementation of GitHub. Can you have a look and let me know if it captures what you had in mind? Although this example only shows how to use it with Kubernetes as a backend (i.e., it will typically be used for on-prem CI testing and building of AI workloads). But If it suits the idea you are thinking about I will clean it up and add the implementations for the other backend as well.

### CODE REVIEW

**Comment by peterschmidt85** [comment]
@Tob-iee I don't think this PR captures what is required.
1. The idea of GitHub action example is to show how to submit dstack runs as a part of GitHub Actions workflows. This implies dstack server is running somewhere else.
1.1. Your PR starts/stops server as a part of the workflow. IMO, this doesn't make sense.
2. The example implies having a page at https://dstack.ai/examples/. Your PR doesn't have it

**Comment by Tob-iee** [comment]
> @Tob-iee I don't think this PR captures what is required.
> 
> 1. The idea of GitHub action example is to show how to submit dstack runs as a part of GitHub Actions workflows. This implies dstack server is running somewhere else.
>    1.1. Your PR starts/stops server as a part of the workflow. IMO, this doesn't make sense.
> 2. The example implies having a page at https://dstack.ai/examples/. Your PR doesn't have it

1. So maybe it's the way that I thought about the example initially. My assumption is that the dstack server runs on where its backend is configured. So typically the idea became that this sample example shows a "continuous deployment" workflow, where the user leverages Dstack in their github actions workflow to automate the deployment of the model on their GPU infrastructure that is on-prem. Which is why I used Kubernetes as the backend.
     1.1. Yes, you are right. I will exclude stopping the server after submitting the deployment run.
2. I will add the page

**Comment by peterschmidt85** [comment]
> My assumption is that the dstack server runs on where its backend is configured. So typically the idea became that this sample example shows a "continuous deployment" workflow, where the user leverages Dstack in their github actions workflow to automate the deployment of the model on their GPU infrastructure that is on-prem. 

Indeed, a dstack server is started where the backend is configured. But, it's not GitHub Actions.

>   1.1. Yes, you are right. I will exclude stopping the server after submitting the deployment run.

Typically, a dstack server is deployed somewhere centrally where the entire team can access it.
Starting server inside GitHub Actions makes no sense.

BTW, in a real scenario (and we can mention that in the example explicitly), the server is running in a private network and if that's the case, GitHub Actions should be configured so that it has access to the private network where the server is running.

> * I will add the page

The page itself is the main asset. Check other pages to understand what level of details and what style is required.
Basically, it should provide minimum but enough information to understand.

**Comment by Tob-iee** [comment]
> > My assumption is that the dstack server runs on where its backend is configured. So typically the idea became that this sample example shows a "continuous deployment" workflow, where the user leverages Dstack in their github actions workflow to automate the deployment of the model on their GPU infrastructure that is on-prem.
> 
> Indeed, a dstack server is started where the backend is configured. But, it's not GitHub Actions.
> 
> > 1.1. Yes, you are right. I will exclude stopping the server after submitting the deployment run.
> 
> Typically, a dstack server is deployed somewhere centrally where the entire team can access it. Starting server inside GitHub Actions makes no sense.
> 
> BTW, in a real scenario (and we can mention that in the example explicitly), the server is running in a private network and if that's the case, GitHub Actions should be configured so that it has access to the private network where the server is running.
> 
> > * I will add the page
> 
> The page itself is the main asset. Check other pages to understand what level of details and what style is required. Basically, it should provide minimum but enough information to understand.

I have added an examples page to help explain the example. However, the summary of the examples use-case is:

You are automating the MLOps CI pipeline using GitHub actions for your in-built MLOps platform or a typical MLOps pipeline. The main idea is that you need your data scientist or ML engineer's ML task, which could be like a fine-tuning, training job, or inference app engine to use a GPU infrastructure because such a job (i.e., AI workload in this case) requires running on a GPU infrastructure because of the expensive computation. Your machine with GPU infrastructure could be on-prem or remote. However, either way, you already have a Dstack server running on the machine to manage and configure your GPU for your AI workloads. So each time the ML engineer or data scientist makes updates to their training component (which could be a script or YAML file) and pushes those changes so that they can execute the AI task, Dstack will automatically run it with the Dstack provisioned GPU infrastructure.

What do you think about the use case/senario?

File: examples/misc/github-actions/.dstack.yml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+          chmod +x ./kind
+          sudo mv ./kind /usr/local/bin/kind
+          echo $PATH
+          which kind
+      
+      - name: Configure Dstack server backend with K8s
+        run: |
+          kind create cluster --config backends/K8s_config/kind_backend_config.yml
+
+      - name: Run Dstack server and capture token
```
Comment (peterschmidt85) [inline examples/misc/github-actions/.dstack.yml]: We agreed that the server is running outside of GitHub Actions

File: examples/misc/github-actions/.dstack.yml [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            exit 1
+          fi
+
+          # Export the token as an environment variable
+          echo "DSTACK_TOKEN=$TOKEN" >> $GITHUB_ENV
+          echo "Captured DSTACK_TOKEN: $TOKEN"
+          
+      - name: Apply run on dstack 
+        run: |
+          git clone https://github.com/dstackai/dstack
```
Comment (peterschmidt85) [inline examples/misc/github-actions/.dstack.yml]: Why do we need some repo if we are already inside our repo with GitHub Actions?

File: examples/misc/github-actions/.dstack.yml [hunk 3]
```diff
@@ -0,0 +1,116 @@
+name: Run dstack Task on GPU
```
Comment (peterschmidt85) [inline examples/misc/github-actions/.dstack.yml]: Also why the file is named .dstack.yml if it's a GitHub Actions file? I suppose GitHub Actions workflows files should be under .github/workflows

File: examples/misc/github-actions/README.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+jobs:
+  run-ai-workload:
+    name: dstack-task
+    runs-on: ubuntu-latest
+
+    steps:
+          - name: Configure SSH to connent to your machine
+            run: |
+              install -m 600 -D /dev/null ~/.ssh/github-actions
```
Comment (peterschmidt85) [inline examples/misc/github-actions/README.md]: What is going on here?

File: examples/misc/github-actions/README.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+              cat >> ~/.ssh/config <<END
+              Host github-actions
+                HostName ${{ secrets.SSH_HOST }}
+                User ${{ secrets.SSH_USER }}
+                IdentityFile ~/.ssh/github-actions
+                Port 22
+                StrictHostKeyChecking no
+              END
+
+          - name: Clone pull new chances of your repo with AI workload code
```
Comment (peterschmidt85) [inline examples/misc/github-actions/README.md]: No idea what is going on here to be honest

**Review by peterschmidt85 — CHANGES_REQUESTED** [review]

File: examples/misc/github-actions/README.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+          Host github-actions
+            HostName ${{ secrets.SSH_HOST }} # The SSH host, provided as a secret
+            User ${{ secrets.SSH_USER }} # The SSH user, provided as a secret
+            IdentityFile ~/.ssh/github-actions # Path to the private key file
+            Port 22 # SSH port (default is 22)
+            StrictHostKeyChecking no # Disable host key checking for simplicity
+          END
+
+      - name: Pull new changes of the repo with AI workload code into your server and run task with dstack
+        run: |
```
Comment (peterschmidt85) [inline examples/misc/github-actions/README.md:57]: What does this step do?

File: examples/misc/github-actions/README.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+              if [ -d "${{ github.repository }}/.git" ]; then
+                cd ${{ github.repository }} && git pull origin $DEFAULT_BRANCH
+              else
+                rm -rf dstack && git clone -b $DEFAULT_BRANCH https://github.com/<"${{ github.repository }}">
+              fi
+            '
+
+      - name: Initialize dstack on the server and run the AI task with dstack
+        run: |
+          ssh github-actions << 'EOF'
```
Comment (peterschmidt85) [inline examples/misc/github-actions/README.md:69]: Why is this needed?

File: examples/misc/github-actions/README.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            # Set environment variables, if any 
+            export <env-variable>=${{ secrets.<env-variable-name> }}
+            export <env-variable>=${{ secrets.<env-variable-name> }}  
+            ....
+            
+            # Activate the Python environment, initialize dstack, and apply the task configuration
+            source <environment-name>/bin/activate
+            cd ${{ github.repository }}
+            dstack init
+            dstack apply -f <path to dstack in your repo directory> --yes &
```
Comment (peterschmidt85) [inline examples/misc/github-actions/README.md:79]: Doesn't look as working code

**Review by peterschmidt85 — CHANGES_REQUESTED** [review]
It seems like there might be some misunderstanding about how GitHub Actions work.

A couple of key points:

There's no need to clone any repositories—GitHub Actions handles that automatically.
Please avoid submitting untested code, whether for this repo or any other.
Given these issues, I have to reject this PR for now. Let me know if you need clarification or guidance—I’m happy to help!
